Cryptonomic/Nautilus-Cloud
# Feature/1 hexagonal architecture

This PR introduces [Hexagonal Architecture](https://blog.octo.com/en/hexagonal-architecture-three-principles-and-an-implementation-example/) into the project. The main goal is to have an application sliced into features instead of classes which are similar from a framework's perspective (like controllers, repositories, ...) and better separation of concerns.

The main assumptions
---
* the core of our application is a `domain` package. We put each service there on the main level. All the services are meant to be our only gateway to our application domain logic. If we need to call a database or other service (via HTTP) we create an interface here. The implementation goes to another package which is...
* adapters package - this is a place for implementation of each interface from domain package, if we need any library (circe, doobie, ...) it should be used only inside an adapter, all parameters and return objects (like DTO's) should be converted inside an adapter to match domain objects we use in interfaces.

How to test it?
---
* since our domain shouldn't have any external dependencies we can test it using unit tests only
* each of our adapters should be tested using integration tests with suitable embedded solution run if possible. 

What next?
---
I want to rebuild Nautilus class in order to be able to run it with a different configuration, therefore we can run simple END-TO-END tests as part of our CI pipeline.
